src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -87,13 +128,97 @@ public override TypedExpression Transform(TypedExpression ex)
             return base.Transform(ex);
         }
 
-        private string CreateTypePrefix(ImmutableDictionary<QsTypeParameter, ResolvedType> typeDict)
+        private Concretion CreateConcretion(ImmutableDictionary<QsTypeParameter, ResolvedType> typeDict)
+        {
+            Concretion result = new Concretion();
+            foreach (var kvp in typeDict)
+            {
+                result.Add((kvp.Key.TypeName, kvp.Value.Resolution));
+            }
+            return result;
+        }
+
+        private Identifier GetConcreteIdentifier(Identifier.GlobalCallable globalCallable, ImmutableDictionary<QsTypeParameter, ResolvedType> types)
+        {
+            Concretion target = CreateConcretion(types);
+
+            // TODO: handle if globalCallable is not in Generics
+
+            var (original, concretions) = Generics[globalCallable.Item];
+
+            NonNullable<string> name;
+
+            if (concretions.ContainsKey(target))
+            {
+                name = concretions[target].FullName.Name;
+            }
+            else
+            {
+                // Create new name
+                string nameString = Guid.NewGuid().ToString() + "_" + globalCallable.Item.Name.Value;
+
+                // TODO: Check for identifier name collisions (not likely)
+                // - check in global namespace
+                // - check in concretion dict for current generic
+
+                name = NonNullable<string>.New(nameString);
+                
+                QsCallable newCallable = ReplaceTypeParamsSyntax.Apply(types, original)
+                    .WithFullName(fullname => new QsQualifiedName(globalCallable.Item.Namespace, name));
+
+                concretions.Add(target, newCallable);
+            }
+
+            return Identifier.NewGlobalCallable(new QsQualifiedName(globalCallable.Item.Namespace, name));
+        }
+    }
+
+    public class ResolveGenericsExpressionKind : ExpressionKindTransformation<ResolveGenericsExpression>
+    {
+        public ResolveGenericsExpressionKind(ResolveGenericsExpression expr) : base(expr) { }
+
+        public override QsExpressionKind<TypedExpression, Identifier, ResolvedType> onIdentifier(Identifier sym, QsNullable<ImmutableArray<ResolvedType>> tArgs)
+        {
+            return base.onIdentifier(sym, tArgs);
```

Comment:
I assume that's preparation for filing something in here? :)

Reply:
Oops, this is actually the start of my work towards making the feature work with partial applications. It will be removed from this PR. Thanks!